src/mono/netcore/System.Private.CoreLib/src/System/Threading/TimerQueue.Browser.cs

```diff
@@ -0,0 +1,127 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Threading
+{
+    // Runtime interface
+    internal static class WasmRuntime
+    {
+        private static Dictionary<int, Action> callbacks = new Dictionary<int, Action>();
+        private static int next_id;
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        private static extern void SetTimeout(int timeout, int id);
+
+        internal static void ScheduleTimeout(int timeout, Action action)
+        {
+            int id = ++next_id;
+            callbacks[id] = action;
+            SetTimeout (timeout, id);
+        }
+
+        // Called by mini-wasm.c:mono_set_timeout_exec
+        private static void TimeoutCallback (int id)
+        {
+            Action cb = callbacks[id];
+            callbacks.Remove(id);
+            cb();
+        }
+    }
+
+    //
+    // WebAssembly-specific implementation of Timer
+    // Based on TimerQueue.Portable.cs
+    //
+    internal partial class TimerQueue
+    {
+        private static List<TimerQueue>? s_scheduledTimers;
+        private static List<TimerQueue>? s_scheduledTimersToFire;
+
+        private static readonly object s_monitor = new object();
+
+        private bool _isScheduled;
+        private long _scheduledDueTimeMs;
+
+        private TimerQueue(int id)
+        {
+        }
+
+        private bool SetTimer(uint actualDuration)
+        {
+            Debug.Assert((int)actualDuration >= 0);
+            long dueTimeMs = TickCount64 + (int)actualDuration;
+            lock (s_monitor)
```

Comment:
This makes me nervous about the lack of thread-safety in WasmRuntime.ScheduleTimeout.